src/creator.rs

```diff
@@ -97,53 +93,244 @@ impl<H: Hasher> Creator<H> {
         }
     }
 
-    fn check_ready(&self) -> bool {
-        if self.current_round == 0 {
-            return true;
-        }
-        if self.current_round > self.max_round {
-            warn!(target: "AlephBFT-creator", "{:?} Maximum round reached. Not creating another unit.", self.node_ix);
-            return false;
+    async fn wait_until_ready(&mut self, round: Round) {
+
+        let prev_round_index = match round.checked_sub(1) {
+            Some(prev_round) => prev_round as usize,
+            None => return,
+        };
+
+        let delay = time::sleep((self.create_lag)(round.into()));
+        tokio::pin!(delay);
+        loop {
+            if usize::from(round) < self.n_candidates_by_round.len() - 1 {
+                return; // Since we get unit from round r, we have enough units from previous rounds to create our unit 
```

Comment:
Two issues:
1) This allows for a very sneaky attack, where an adversarial node does not follow the schedule and creates unit right away -- all other nodes will follow him in this and as a consequence, they will reach MAX_ROUND early and crash. This can be fixed by requiring that the round is higher by one, i.e. the condition should be:
`usize::from(round) + 2 < self.n_candidates_by_round.len()`
(Note we cannot subtract 2 from the RHS because of unsigned variable).
2) The comment is inaccurate: the point is not that we can create a unit, the point is that we can skip the delay, because we are behind.

Reply:
Very nice catch with 1., I would add that we want to have a comment explaining this specific attack there, because the `2` in the condition will certainly look curious.

Reply:
Done